k2/csrc/array.h

```diff
@@ -52,70 +56,150 @@ struct Array2Size {
   // `size1` is the top-level size of the array, equal to the object's .size
   // element
   I size1;
-  // `size2` is the nunber of elements in the array, equal to
+  // `size2` is the number of elements in the array, equal to
   // o->indexes[o->size] - o->indexes[0] (if the Array2 object o is
   // initialized).
   I size2;
 };
 
-template <typename I, typename Ptr>
+template <typename Ptr, typename I = int32_t>
 struct Array2 {
   // Irregular two dimensional array of something, like vector<vector<X> >
   // where Ptr is, or behaves like, X*.
   using IndexT = I;
   using PtrT = Ptr;
+  using ValueType = typename std::iterator_traits<Ptr>::value_type;
 
   IndexT size;
-  const IndexT *indexes;  // indexes[0,1,...size] should be defined; note, this
-                          // means the array must be of at least size+1.  We
-                          // require that indexes[i] <= indexes[i+1], but it is
-                          // not required that indexes[0] == 0, it may be
-                          // greater than 0.
+  IndexT *indexes;  // indexes[0,1,...size] should be defined; note, this
+                    // means the array must be of at least size+1.  We
+                    // require that indexes[i] <= indexes[i+1], but it is
+                    // not required that indexes[0] == 0, it may be
+                    // greater than 0.
 
   PtrT data;  // `data` might be an actual pointer, or might be some object
               // supporting operator [].  data[indexes[0]] through
               // data[indexes[size] - 1] must be accessible through this
               // object.
 
-  /* initialized definition:
-
-        An Array2 object is initialized if its `size` member is set and its
-        `indexes` and `data` pointer allocated, and the values of its `indexes`
-        array are set for indexes[0] and indexes[size].
+  /*
+        An Array2 object is non-empty if its `size` member is set and its
+        `indexes` and `data` pointer allocated,
   */
+  bool Empty() const {
+    return size == 0 || indexes == nullptr || data == nullptr;
```

Comment:
The 'initialized' definition was actually needed, and an Array2 object can be initialized if it has size == 0 and indexes== pointer to e.g. [0], and data == nullptr.  It would just be empty in the sense of no elements.  The point of `initialized` is that it is set up.
The reason for having a definition is that some functions may require the Array2 object to be initialized (and some not).
Maybe just keep that comment?
Empty() would just return {size == 0}.

Reply:
OK, will get back the comment and update Empty()